allennlp/tests/semparse/domain_languages/quarel_language_test.py

```diff
@@ -0,0 +1,127 @@
+                 'World -> world1']
+
+    def test_infer_quaval(self):
+        assert self.language.execute(('(infer (and (thickness low world1) '
+                                      '(thickness high world2)) '
+                                      '(strength lower world1) '
+                                      '(strength lower world2))')) == 0
+
+        assert self.language.execute(('(infer (and (thickness low world1) '
+                                      '(thickness high world2)) '
+                                      '(strength lower world2) '
+                                      '(strength lower world1))')) == 1
+
+        assert self.language.logical_form_to_action_sequence(('(infer (and (thickness low world1) '
+                                                              '(thickness high world2)) '
+                                                              '(strength lower world1) '
+                                                              '(strength lower world2))')) == \
+                ['@start@ -> int',
+                 'int -> [<QuaRelType,QuaRelType,QuaRelType:int>, QuaRelType, QuaRelType, '
+                 'QuaRelType]',
+                 '<QuaRelType,QuaRelType,QuaRelType:int> -> infer',
+                 'QuaRelType -> [<QuaRelType,QuaRelType:QuaRelType>, QuaRelType, QuaRelType]',
+                 '<QuaRelType,QuaRelType:QuaRelType> -> and',
+                 'QuaRelType -> [<Direction,World:QuaRelType>, Direction, World]',
+                 '<Direction,World:QuaRelType> -> thickness',
+                 'Direction -> low',
+                 'World -> world1',
+                 'QuaRelType -> [<Direction,World:QuaRelType>, Direction, World]',
+                 '<Direction,World:QuaRelType> -> thickness',
+                 'Direction -> high',
+                 'World -> world2',
+                 'QuaRelType -> [<Direction,World:QuaRelType>, Direction, World]',
+                 '<Direction,World:QuaRelType> -> strength',
+                 'Direction -> lower',
+                 'World -> world1',
+                 'QuaRelType -> [<Direction,World:QuaRelType>, Direction, World]',
+                 '<Direction,World:QuaRelType> -> strength',
+                 'Direction -> lower',
+                 'World -> world2']
+
+    def test_action_sequence_to_logical_form(self):
+        logical_form = ('(infer (and (thickness low world1) '
+                        '(thickness high world2)) '
+                        '(strength lower world1) '
+                        '(strength lower world2))')
+        action_sequence = self.language.logical_form_to_action_sequence(logical_form)
+        recovered_logical_form = self.language.action_sequence_to_logical_form(action_sequence)
+        assert recovered_logical_form == logical_form
+
+    def test_get_valid_actions(self):
+        valid_actions = self.language.get_valid_actions()
+        assert set(valid_actions.keys()) == {
+                'QuaRelType',
```

Comment:
I think it's good practice to test the productions for all of these keys.  I see 8 keys in here and only 6 `check_productions_match` calls.  It's also nice if these are sorted in some reasonable way, and if the order of the `checks` below matches, so you can easily find things.